crates/gpui/src/platform/windows/window.rs

```diff
@@ -109,75 +143,421 @@ impl WindowsWindowInner {
+            VK_NEXT => Some(("pagedown", None)),
+            VK_ESCAPE => Some(("escape", None)),
+            VK_INSERT => Some(("insert", None)),
+            _ => None,
+        };
+
+        if let Some((key, ime_key)) = key {
+            Some(Keystroke {
+                modifiers: self.current_modifiers(),
+                key: key.to_string(),
+                ime_key: ime_key.map(|k| k.to_string()),
+            })
+        } else {
+            None
+        }
+    }
+
+    fn handle_keydown_msg(&self, wparam: WPARAM) -> CallbackResult {
+        let mut callbacks = self.callbacks.borrow_mut();
+        let keystroke = self.parse_key_msg_keystroke(wparam);
+        if let Some(keystroke) = keystroke {
+            if let Some(callback) = callbacks.input.as_mut() {
+                let event = KeyDownEvent {
+                    keystroke,
+                    is_held: true,
+                };
+
+                if callback(PlatformInput::KeyDown(event.clone())) {
+                    if let Some(request_frame) = callbacks.request_frame.as_mut() {
+                        request_frame();
                     }
+                    CallbackResult::Handled { by_callback: true }
+                } else if let Some(mut input_handler) = self.input_handler.take() {
+                    if let Some(ime_key) = &event.keystroke.ime_key {
+                        input_handler.replace_text_in_range(None, ime_key);
+                    }
+                    self.input_handler.set(Some(input_handler));
+                    if let Some(request_frame) = callbacks.request_frame.as_mut() {
+                        request_frame();
+                    }
+                    CallbackResult::Handled { by_callback: true }
+                } else {
+                    CallbackResult::Handled { by_callback: false }
+                }
+            } else {
+                CallbackResult::Handled { by_callback: false }
+            }
+        } else {
+            CallbackResult::Unhandled
+        }
+    }
+
+    fn handle_keyup_msg(&self, wparam: WPARAM) -> CallbackResult {
+        let mut callbacks = self.callbacks.borrow_mut();
+        let keystroke = self.parse_key_msg_keystroke(wparam);
+        if let Some(keystroke) = keystroke {
+            if let Some(callback) = callbacks.input.as_mut() {
+                let event = KeyUpEvent { keystroke };
+                CallbackResult::Handled {
+                    by_callback: callback(PlatformInput::KeyUp(event)),
+                }
+            } else {
+                CallbackResult::Handled { by_callback: false }
+            }
+        } else {
+            CallbackResult::Unhandled
+        }
+    }
+
+    fn handle_char_msg(&self, wparam: WPARAM) -> LRESULT {
+        let mut callbacks = self.callbacks.borrow_mut();
+        if let Some(callback) = callbacks.input.as_mut() {
+            let modifiers = self.current_modifiers();
+            let msg_char = wparam.0 as u8 as char;
```

Comment:
https://learn.microsoft.com/en-us/windows/win32/inputdev/wm-char
Since `WM_CHAR` accepts UTF-16 code units, this conversion can cause problems.

Reply:
I agree this is an issue, but I think it should maybe be covered in another PR. I only say this because the char encoding situation is a little wack on windows. Recently they started emphasizing UTF-8 with the `RegisterClassA` call even tho previously they said new applications should use `RegisterClassW`.

https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page
> Until recently, Windows has emphasized "Unicode" -W variants over -A APIs. However, recent releases have used the ANSI code page and -A APIs as a means to introduce UTF-8 support to apps. If the ANSI code page is configured for UTF-8, then -A APIs typically operate in UTF-8. This model has the benefit of supporting existing code built with -A APIs without any code changes.

That being said I could change this to what this PR does if this is a blocker: https://github.com/zed-industries/zed/pull/8809/files#diff-f5cac42313d3717a63ced25b77b50370494328242025c977fa77d17d5ad1bb40R192-R193
```rs
let Ok(first_char) = char::decode_utf16(src).collect::<Vec<_>>()[0] else {
    return None;
};
```

Reply:
Oh I see. Then we should use `-A` APIs